lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,506 @@
+
+/*
+ * @see PHP docs for array_is_list: user-contributed polyfill
+ */
+if ( ! function_exists( 'array_is_list' ) ) {
+	function array_is_list( $array ) {
+		$i = 0;
+
+		foreach ( $array as $k => $v ) {
+			if ( $k !== $i++ ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+}
+
+class Selector {
+	public $element = null;
+	public $class = null;
+	public $hash = null;
+	public $has_attribute = null;
+	public $then = null;
+}
+
+class ParseResult {
+	public $last_index;
+	public $selector = null;
+}
+
+class WP_HTML_Attribute_Sourcer {
+	/**
+	 * Attributes definitions, typically from `block.json`.
+	 *
+	 * @see WP_Block_Type_Registry
+	 *
+	 * @var mixed|null
+	 */
+	public $attribute_definitions;
+
+	/**
+	 * Source HTML containing embedded attributes.
+	 *
+	 * @var mixed|null
+	 */
+	private $html;
+
+	public function __construct( $attribute_definitions = null, $html = null ) {
+		$this->attribute_definitions = $attribute_definitions;
+		$this->html = $html;
+	}
+
+	public function source_attributes() {
+		$attributes = [];
+		$unparsed = [];
+
+		foreach ( $this->attribute_definitions as $name => $definition ) {
+			$sourcer = self::parse_definition( $definition );
+			switch ( $sourcer ) {
+				case null:
+				case 'not-sourced':
+				case 'unsupported':
+					$unparsed[] = $name;
+					continue 2;
+
+				case 'inner-html':
+					$attributes[ $name ] = $this->html;
+					continue 2;
+			}
+
+			$tags = self::select( $sourcer['selector'], $this->html );
+			if ( null === $tags ) {
+				$unparsed[] = $name;
+				continue;
+			}
+
+			switch ( $sourcer['type'] ) {
+				case 'html':
+					$attributes[ $name ] = $tags->get_content_inside_balanced_tags();
+					continue 2;
+
+				case 'attribute':
+					$attributes[ $name ] = $tags->get_attribute( $sourcer['attribute'] );
+					continue 2;
+			}
+		}
+
+		return array(
+			'attributes' => $attributes,
+			'unparsed' => $unparsed
+		);
+	}
+
+	public static function select_match( $tags, $s ) {
+		if ( ! empty( $s['tag_name'] ) && strtoupper( $s['tag_name'] ) !== $tags->get_tag() ) {
+			return null;
+		}
+
+		if ( ! empty( $s['class_names'] ) ) {
+			$classes = $tags->get_attribute( 'class' );
+			if ( null === $classes ) {
+				return null;
+			}
+
+			foreach ( $s['class_names'] as $class_name ) {
+				if ( ! preg_match( "~\b{$class_name}\b~", $classes ) ) {
+					return null;
+				}
+			}
+		}
+
+		if ( isset( $s['hash'] ) && $s['identifier'] !== $tags->get_attribute( 'id' ) ) {
+			return null;
+		}
+
+		if ( isset( $s['has_attribute'] ) && null === $tags->get_attribute( $s['has_attribute'] ) ) {
+			return null;
+		}
+
+		return $tags;
+	}
+
+	public static function select( $selectors, $html ) {
+		$tags = new WP_HTML_Processor( $html );
+		if ( ! $tags->next_tag() ) {
+			return null;
+		}
+
+		$tags->set_bookmark( 'start' );
+
+		foreach ( $selectors as $s ) {
+			$tags->seek( 'start' );
+			$max = 100;
+			while ( --$max > 0 ) {
+				$next = $s;
```

Comment:
Here's an idea – take a selector line `#main > .wrapper a span`.

This class is now going top to bottom and matching `#main`, then the immediate descendant `.wrapper`, then a descendant `a` and then a descendant `span`.

What if it went bottom to top? So found the first `span` in the document, and then investigated the parents stack to see if it matches the selector. Now we don't need to jump around the document to decide if it's a match.

There's still the issue of `~` and `+` combinators, but we could keep track of a tree of relevant siblings. For example in a HTML like

```html
<div id="main">
   <div class="wrapper">
     <a>
       <div></div>
     </a>
     <a>
         <span></span>
         <span></span> <--- the cursor is here
     </a>
  </div>
</div>
```

We'd only store the following data structure (without the children of the initial `a`)

```
div#main
   div.wrapper
      a
      a
         span
         span
```

If we want to support matching arbitrary attributes beyond ids and classes, we could keep a bookmark for each parent and backtrack if needed.

Reply:
[That's what the browsers do, by the way](https://stackoverflow.com/questions/5797014/why-do-browsers-match-css-selectors-from-right-to-left).

Reply:
Yeah this seems worth a try. If we aren't going to fully parse the HTML then I don't see the same value add because _quickly eliminating potential matches_ would happen before diving deep into the structure (and maybe that can still happen). If on the other hand, we're going to do a full document parse (and carry the weight of all that implies) then this method seems approachable.

The one major difference I see at this point is that if we're looking for the inner-most element first then we have to run `$tags->matches()` on every node and run a full check on every node. In the approach I've started taking we prune out most nodes out the door and are able to iterate with less careful checking, because at that point we're mostly just checking tag name and whether it's an opener or closer. In practice it may not matter, but I did want to point out something different from our HTML-based approach vs. what might be happening in the DOM land in the browser.